### PR TITLE
Refactor `.gitattributes` and `.gitignore`

### PR DESCRIPTION
**Pull request by 1Solon**
### Description

This PR splits the global `.gitignore` into local `.gitignore`s and refactors the `.gitattributes` file.

### Type of change

Please select the option that best describes the changes made:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Changes

- **Splits global `.gitignore` into local `.gitignore`s**
- **Adds fully fleshed out `.gitattributes`**

### Rationale for a `breaking change`

Certain files *(for example jupyter notebooks)* are encoded with `crlf` currently, when this change is merged- those files when touched will immediately change to the current systems line endings. This is desired. Committing the file will resolve this issue. 

### Rationale for changing `.gitignore`

Our repository is a monorepo- in the beginning I felt that having a central `.gitignore` would offer a single point of call for any changes that needed to be made to it. However, in general this has only really caused confusion and made the repository more difficult to manage.

By splitting the `.gitignore` into local `.gitignore`s, we can now have a `.gitignore` in each directory that is specific to that directory. Hopefully allowing individual pairs to define their own `.gitignore` rules.

### Rationale for changing `.gitattributes`

Windows and linux users use different [newline](https://en.wikipedia.org/wiki/Newline) characters, in git repositories with multiple users using different systems. This can cause problems. In particular, we have users in this repository who use both Windows and MacOS. An example of what may happen is below:

Let's say I have two text files, each created by a different user. If I run `git diff` on these files, I'll get the below result:

```diff
diff --git a/file.txt b/file.txt
index 83db48f..7df2d21 100644
--- a/file.txt
+++ b/file.txt
@@ -1,4 +1,4 @@
-This is a line of text in Linux (LF)
+This is a line of text in Windows (CRLF)
```

As you can see, the only difference between the two files is the newline character. This can cause a lot of noise in the git history and make it difficult to see what changes have actually been made.

The solution to this is setting line-ending rules. You can achieve this through `.gitattributes`.

Previously, our `.gitattributes` looked like this:

```python
# Apply CRLF to all files
* text eol=crlf

# Exclude .zip files from EOL normalization
*.zip -text

# Exlucde .jpg files from EOL normalization
*.jpg -text

# Exlucde .png files from EOL normalization
*.png -text
```

As you can see, I'm taking a very heavy handed approach to this problem. I'm applying `crlf` to all files, and then excluding other files as they become a problem. As recent branch-switching issues have shown, this is not a sustainable solution.

I've now refactored the `.gitattributes` to look like this:

```python
# Apply CRLF to Go files
*.go text eol=crlf
*.mod text eol=crlf
*.sum text eol=crlf

# Apply CRLF to SQL files
*.sql text eol=crlf

# Apply CRLF to LaTeX files
*.tex text eol=crlf
*.bib text eol=crlf

# Appky CRLF to Python files
*.py text eol=crlf

# Apply CRLF to Markdown files
*.md text eol=crlf

# Apply CRLF to text files
*.txt text eol=crlf

# Apply CRLF to Yaml files
*.yaml text eol=crlf
*.yml text eol=crlf

# Apply CRLF to configuration files
*.conf text eol=crlf
*.cfg text eol=crlf
*.mjs text eol=crlf

# Apply CRLF to TypeScript files
*.ts text eol=crlf
*.tsx text eol=crlf

# Apply CRLF to JavaScript files
*.js text eol=crlf

# Apply CRLF to CSS files
*.css text eol=crlf
```

This is a much more sustainable solution. I'm now applying `crlf` to specific file types, rather than all files. This should solve the problem, while also preventing the corruption of specific binary files we've been seeing.

### CODE REVIEW

File: .gitattributes
Comment (ankraus) [inline .gitattributes:1]: Good idea to reverse that policy, especially since we had so many issues with binary files

File: Backend/.gitignore
Comment (ankraus) [inline Backend/.gitignore:1]: No concerns from my side regarding this, all necessary ignores for the backend are covered

**Review by ankraus — APPROVED** [review]
Great work on this PR, well explained!

**Review by anaisbl — APPROVED** [review]

**Review by JessicaFornetti — APPROVED** [review]

**Review by KaustubhTrivedi — APPROVED** [review]

**Comment by KaustubhTrivedi** [comment]
Looks good on my part.

**Comment by KaustubhTrivedi** [comment]
> Great work on this PR, well explained!

+1

**Pull request merged**